### PR TITLE
[VPU] Fix batch detection/removal logic in MYX plugin

### PR DESCRIPTION
**Pull request by KorA6**
### Details:
 - *This PR is a duplicate of #3143 with some code style changes*

### Tickets:
 - *50877*

### CODE REVIEW

File: inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp [hunk 1]
```diff
@@ -82,94 +63,66 @@ void FrontEnd::detectNetworkBatch(
...
-            }
-
-            // 2. Don't support if there multiple outputs as well
-            if (outputsInfo.size() != 1) {
-                VPU_THROW_FORMAT("Unsupported network configuration : layer %s must be the only output of the network", layer->get_name());
-            }
-            model->attrs().set<bool>("withDetectionOutput", true);
+        // Don't support if there multiple outputs as well
+        if (outputsInfo.size() != 1) {
+            env.log->trace("Unsupported network configuration : layer %s must be the only output of the network", layer->get_name());
```
Comment (Maxim-Doronin) [inline inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp:107]: Please, also mention operation type info

Comment (Maxim-Doronin) [reply inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp:107]: Why `env.log->trace` instead of `VPU_THROW_UNLESS`?

File: inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp [hunk 2]
```diff
@@ -82,94 +63,66 @@ void FrontEnd::detectNetworkBatch(
...
-        for (const auto& layer : layers) {
-            if (layer->get_type_name() != std::string("DetectionOutput"))
-                continue;
-
-            env.log->trace("Found DetectionOutput layer [%s]", layer->get_name());
-
-            if (layer->get_output_size() == 0)
-                VPU_THROW_FORMAT("Unsupported layer %s configuration: no outputs", layer->get_name());
-
-            // 1. Don't support if DetectionOutput is not the last layer in network
```
Comment (Maxim-Doronin) [inline inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp:143]: Why did you remove this check?

File: inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp [hunk 3]
```diff
@@ -82,94 +63,66 @@ void FrontEnd::detectNetworkBatch(
 
         auto ieShapes = ieData->getTensorDesc().getDims();
         env.log->trace("Input [%s] : %v", p.first, ieShapes);
+        // assume only 4D and 5D inputs have batch
+        if (ieShapes.size() == 4 || ieShapes.size() == 5) {
```
Comment (Maxim-Doronin) [inline inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp:67]: Should we remove `DETECT_NETWORK_BATCH` config from tests when testing NC layouts?

**Review by Maxim-Doronin — COMMENTED** [review]
Are you going to add any tests on the new behavior?

File: inference-engine/src/vpu/graph_transformer/src/backend/serialize.cpp [hunk 4]
```diff
@@ -147,7 +147,10 @@ void BackEnd::serializeConstShapes(const Model& model, const mv_blob_header& blo
             serializeToBlob(data->desc().dims(), shapeLocation.dimsOffset);
         } else if (data->usage() == DataUsage::Output || data->usage() == DataUsage::Input) {
             auto ioDimsUpperBoundOffset = data->attrs().get<int>("ioDimsUpperBoundOffset");
-            serializeToBlob(data->desc().dims(), ioDimsUpperBoundOffset);
+            DimValues d = data->desc().dims();
```
[the remark this answers is not in the document]
Comment (KorA6) [reply inference-engine/src/vpu/graph_transformer/src/backend/serialize.cpp]: Changed

File: inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp (hunk 3 above)
Comment (KorA6) [reply inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp:67]: I don't think so.
In the case with NC layout, we just call `return` on line 89.

File: inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp [hunk 5]
```diff
@@ -82,94 +63,66 @@ void FrontEnd::detectNetworkBatch(
...
-            if (layer->outData.empty()) {
-                VPU_THROW_FORMAT("Unsupported layer %s configuration: no outputs", layer->name);
-            }
-
-            // 1. Don't support if DetectionOutput is not the last layer in network
-            if (!ie::getInputTo(layer->outData.front()).empty()) {
-                VPU_THROW_FORMAT("Unsupported layer %s configuration : it is not a network output", layer->name);
-            }
+    const auto layers = network.getFunction()->get_ordered_ops();
+    for (const auto& layer : layers) {
```
[the remark this answers is not in the document]
Comment (KorA6) [reply inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp]: Changed.

File: inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp (hunk 1 above)
Comment (KorA6) [reply inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp]: Type info was added.
The error message was changed.

File: inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp (hunk 2 above)
Comment (KorA6) [reply inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp:143]: This check has been replaced by 
```
VPU_THROW_UNLESS(op->get_output_size() == 1, "Unsupported layer {} with type {} must have only 1 output but {} provided", op->get_name(), op->get_type_name(), op->get_output_size());
```

**Comment by KorA6** [comment]
> Are you going to add any tests on the new behavior?

I'm working on it. The test will be added to this PR soon.

File: inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp [hunk 6]
```diff
@@ -59,117 +44,91 @@ void FrontEnd::detectNetworkBatch(
...
+    size_t dimsSize = 0;
+    ShapesMap inputShapes;
+    
+    for (const auto& p : inputsInfo) {
+        const auto info = p.second;
+        const auto ieData = info->getInputData();
+        auto ieShapes = ieData->getTensorDesc().getDims();
+        if (!dimsSize)
+            dimsSize = ieShapes.size();
+        if (dimsSize != ieShapes.size())
```
Comment (Maxim-Doronin) [inline inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp:96]: Do you accept only inputs with the same number of dims? Why? 
What about a case when the first input is 4D and the second is 2D img_info?

File: inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp [hunk 7]
```diff
@@ -59,117 +44,86 @@ void FrontEnd::detectNetworkBatch(
...
-        default:
-            VPU_THROW_FORMAT("Input %v has unexpected layout %v", ieData->getName(), ieData->getLayout());
+        // assume only 4D and 5D inputs have batch
+        if (ieShapes.size() == 4 || ieShapes.size() == 5) {
+            auto batch = ieShapes[0];
+            if (!inputBatch)
+                inputBatch = batch;
+
+            if (inputBatch != batch) {
+                env.log->trace("Network input with name %s has different batch dim then previous inputs, aborting batch detection", p.first);
```
Comment (Maxim-Doronin) [inline inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp:71]: then -> compared to

File: inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp [hunk 8]
```diff
@@ -59,117 +44,86 @@ void FrontEnd::detectNetworkBatch(
...
+        return;
+    }
 
-            env.log->trace("Found DetectionOutput layer [%s]", layer->name);
+    ShapesMap inputShapes;
 
-            if (layer->outData.empty()) {
-                VPU_THROW_FORMAT("Unsupported layer %s configuration: no outputs", layer->name);
-            }
+    for (const auto& p : inputsInfo) {
```
Comment (Maxim-Doronin) [inline inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp:89]: style: p -> inputInfo

File: inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp [hunk 9]
```diff
@@ -59,117 +44,86 @@ void FrontEnd::detectNetworkBatch(
...
+        const auto info = p.second;
+        const auto ieData = info->getInputData();
+        auto ieShapes = ieData->getTensorDesc().getDims();
+        ieShapes[0] = 1;
```
Comment (Maxim-Doronin) [inline inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp:93]: Please, also check that `ieShapes[0] == inputBatch` before assigning 1

File: inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp (hunk 2 above)
Comment (Maxim-Doronin) [reply inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp:143]: It checks a number of layer outputs, not network outputs, doesn't it?

File: inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp [hunk 10]
```diff
@@ -202,6 +156,7 @@ void FrontEnd::detectNetworkBatch(
 
             _unbatchedOutputs.insert(ieData);
         }
+
         // preserve overriden output precision.
         ieData->setPrecision(outputsPresisions[p.first]);
```
Comment (Maxim-Doronin) [inline inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp:161]: Looks suspecting. Do we need it? But I suggest taking a look at it during 48878. Just keep in mind

File: inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp [hunk 11]
```diff
@@ -202,6 +156,7 @@ void FrontEnd::detectNetworkBatch(
 
             _unbatchedOutputs.insert(ieData);
```
Comment (Maxim-Doronin) [inline inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp:157]: Could you find any real cases when `origShape == newShape` is true? It looks redundant to me.

File: inference-engine/tests/functional/plugin/myriad/shared_tests_instances/behavior/detect_network_batch_test.cpp [hunk 12]
```diff
@@ -0,0 +1,22 @@
+// Copyright (C) 2018-2021 Intel Corporation
+// Copyright (C) 2018-2021 Intel Corporation
```
Comment (adudchen) [inline inference-engine/tests/functional/plugin/myriad/shared_tests_instances/behavior/detect_network_batch_test.cpp:2]: Duplicate copyright

File: inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp [hunk 13]
```diff
@@ -196,12 +154,12 @@ void FrontEnd::detectNetworkBatch(
 
         const auto origShape = outputShapes[p.first];
         const auto newShape = ieData->getDims();
-
+        VPU_THROW_UNLESS(origShape != newShape, "Have unbatched output");
         if (origShape == newShape) {
```
Comment (adudchen) [inline inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp:158]: I think this check is not required, given that we are checking the inequality at line 157

Comment (KorA6) [reply inference-engine/src/vpu/graph_transformer/src/frontend/detect_network_batch.cpp:158]: This line was added for trying to find real use cases by using dlbenchmark.
The PR will have to be reworked again after the validation.

File: inference-engine/tests/functional/plugin/myriad/shared_tests_instances/behavior/detect_network_batch_test.cpp (hunk 12 above)
Comment (KorA6) [reply inference-engine/tests/functional/plugin/myriad/shared_tests_instances/behavior/detect_network_batch_test.cpp]: Changed.

**Review by Maxim-Doronin — APPROVED** [review]

**Review by adudchen — APPROVED** [review]
